src/passes/OptimizeInstructions.cpp

```diff
@@ -933,45 +933,51 @@ struct OptimizeInstructions
       return;
     }
     optimizeMemoryAccess(curr->ptr, curr->offset);
-    if (curr->valueType.isInteger()) {
-      // truncates constant values during stores
-      // (i32|i64).store(8|16|32)(p, C)   ==>
-      //    (i32|i64).store(8|16|32)(p, C & mask)
-      if (auto* c = curr->value->dynCast<Const>()) {
-        if (curr->valueType == Type::i64 && curr->bytes == 4) {
-          c->value = c->value.and_(Literal(uint64_t(0xffffffff)));
-        } else {
-          c->value = c->value.and_(Literal::makeFromInt32(
-            Bits::lowBitMask(curr->bytes * 8), curr->valueType));
-        }
+    optimizeStoredValue(curr->value, curr->bytes);
+    if (auto* unary = curr->value->dynCast<Unary>()) {
+      if (unary->op == WrapInt64 && curr->bytes == 4) {
+        // instead of wrapping to 32, just store some of the bits in the i64
+        curr->valueType = Type::i64;
+        curr->value = unary->value;
+      }
+    }
+  }
+
+  void optimizeStoredValue(Expression*& value, Index bytes) {
+    if (!value->type.isInteger()) {
+      return;
+    }
+    // truncates constant values during stores
+    // (i32|i64).store(8|16|32)(p, C)   ==>
+    //    (i32|i64).store(8|16|32)(p, C & mask)
+    if (auto* c = value->dynCast<Const>()) {
+      if (value->type == Type::i64 && bytes == 4) {
+        c->value = c->value.and_(Literal(uint64_t(0xffffffff)));
+      } else {
+        c->value = c->value.and_(
+          Literal::makeFromInt32(Bits::lowBitMask(bytes * 8), value->type));
       }
     }
     // stores of fewer bits truncates anyhow
-    if (auto* binary = curr->value->dynCast<Binary>()) {
+    if (auto* binary = value->dynCast<Binary>()) {
       if (binary->op == AndInt32) {
         if (auto* right = binary->right->dynCast<Const>()) {
           if (right->type == Type::i32) {
```

Comment:
This section could use a matcher if you wanted.

Reply:
This could be refactored to be shorter, good point. I'll leave that for a separate PR though, I tried to modify the existing code as little as possible.